remove hard-coded /admin/ path

gstrauss opened the pull request:
**What does this PR aim to accomplish?:**

remove hard-coded /admin/ path; relocatable

This does not change or require any changes to where pi-hole installs AdminLTE, which at the moment is `/var/www/html/admin/`, and accessed via `http://pi.hole/admin/`, but this change allows an administrator to manually place the files elsewhere, and to allow the same functionality when accessed from an alternative url-path, e.g. `http://localhost/pihole/admin/`, a url-path less likely to conflict with an already existing website.

This PR uses a suffix match for `/login.php` rather than exact match `/admin/login.php`

---
**By submitting this pull request, I confirm the following:**

1. I have read and understood the [contributors guide](https://docs.pi-hole.net/guides/github/contributing/), as well as this entire template. I understand which branch to base my commits and Pull Requests against.
2. I have commented my proposed changes within the code and I have tested my changes.
3. I am willing to help maintain this change if there are issues with it later.
4. It is compatible with the [EUPL 1.2 license](https://opensource.org/licenses/EUPL-1.1)
5. I have squashed any insignificant commits. ([`git rebase`](http://gitready.com/advanced/2009/02/10/squashing-commits-with-rebase.html))
6. I have checked that another pull request for this purpose does not exist.
7. I have considered, and confirmed that this submission will be valuable to others.
8. I accept that this submission may not be used, and the pull request closed at the will of the maintainer.
9. I give this submission freely, and claim no ownership to its content.

---
- [x] I have read the above and my PR is ready for review. *Check this box to confirm*

gstrauss committed:
remove hard-coded /admin/ path; relocatable

Signed-off-by: Glenn Strauss <gstrauss@gluelogic.com>

gstrauss commented:
@rdwebdesign I think this PR is up your alley.

Is this error perhaps due to recent changes elsewhere which broke CI for the pi-hole repo?
```
Run sudo update-alternatives --set php /usr/bin/php8.0
update-alternatives: error: alternative /usr/bin/php8.0 for php not registered; not setting
Error: Process completed with exit code 2.
```

gstrauss commented:
Where can I see the errors during linting from my changes to password.php?
```
Files that were not fixed due to errors reported during linting before fixing:
   1) /github/workspace/scripts/pi-hole/php/password.php
```

rdwebdesign commented:
> Is this error perhaps due to recent changes elsewhere which broke CI for the pi-hole repo?

This error was already fixed on development branch (`devel`).

You need to base your code (and target your PR) on `devel` branch.

gstrauss commented:
> > Is this error perhaps due to recent changes elsewhere which broke CI for the pi-hole repo?
> 
> This error was already fixed on development branch (`devel`).
> 
> You need to base your code (and target your PR) on `devel` branch.

Okay.  Updated PR and rebased on `devel`.

(The `devel` branch here is named differently from `development` in the other repos.  Close, but not the same name!)

rdwebdesign commented on scripts/pi-hole/php/password.php:
```diff
@@ -68,9 +68,15 @@ function verifyPassword($pwhash, $use_api = false)
+                        ? substr_compare($script_name, '/login.php', -10)
+                        : $script_name === '/login.php') {
+                        header('Location: '.$redirect_url);
+                        exit;
```
I think this code will work for any `$_SERVER['SCRIPT_NAME']` length:

```suggestion
                if ($_SERVER['REQUEST_METHOD'] === 'POST' && substr_compare($_SERVER['SCRIPT_NAME'], '/login.php', -10) === 0) {
                    header('Location: '.$redirect_url);
                    exit;
```

rdwebdesign replied:
OK.
I think my suggestion might generate warnings on PHP versions older than 7.2.18 (I'm unable to test it right now).

If that's the case, what do you think about this?

```suggestion
                if (
                    $_SERVER['REQUEST_METHOD'] === 'POST' &&
                    strlen($_SERVER['SCRIPT_NAME']) >= 10 &&
                    substr_compare($_SERVER['SCRIPT_NAME'], '/login.php', -10) === 0
                ) {
                    header('Location: '.$redirect_url);
                    exit;
```

gstrauss replied:
> OK. I think my suggestion might generate warnings on PHP versions older than 7.2.18 (I'm unable to test it right now).
> 
> If that's the case, what do you think about this?
[snip]

What I think: I am confused.  Are you reviewing my code, which should work portably?  Do you know why my code might not work portably?  Could you please give me feedback on that?  If you are suggesting changes, would you please help me to understand why the changes you are suggesting are better, and what issues you are trying to solve by suggesting changes?  Is it code syntax?  code formatting?  Something else?

I am unable to evaluate your suggestions without knowing more.  There are some potential limitations in [substr_compare()](https://www.php.net/manual/en/function.substr-compare.php) in older versions of PHP, which might generate a warning for certain string lengths of `haystack`, the first argument.  That is why my code uses a ternary expression to handle those cases without issuing a warning.   I have not tested older PHP versions with `needle` longer than `haystack`, but my code is paranoid to avoid doing so.  Maybe the paranoia is unnecessary?  Even if it is too paranoid, does this need to be optimized?  

Note: there is a bug in the code which I will fix momentarily: `substr_compare() === 0` must be the test for success.

gstrauss replied:
To hopefully shortcut some back-and-forth: PHP is not my personal language of choice.  If you tell me to "change the code to X", then I will change the code to X.

rdwebdesign replied:
I'm not trying to impose my code or trying to "teach" you anything.

I suggested this change because the use of the ternary operator inside an expression for `if` is confusing to understand and can lead to "hard to find bugs".

My suggested code tests the haystack length before calling `substr_compare()`. 
If the length test fails, `substr_compare()` will never be executed and there will be no warnings.

You can use whatever you like. I just suggested a different one for readability (and it adds the missing `=== 0`).

gstrauss replied:
> I suggested this change because the use of the ternary operator inside an expression for if is confusing to understand and can lead to "hard to find bugs".

That may be more true for nested ternary operations (`a < b ? b < c ? c  : b : a`).  However, wrapping long lines of code into multiple lines is very similar in "complexity" when using `&&` or one `?:`.

Since you're the one doing the review, I'll implement your suggestion (including inefficiently performing 3x the lookup of `$_SERVER['SCRIPT_NAME']`)

gstrauss commented:
It would appear that stickler-ci part of the tests is hung or extremely backlogged.
That test has been queued for almost 30 mins.

rdwebdesign commented:
Reload the page.
It's probably a problem with the web interface.

The tests were finished:
![image](https://user-images.githubusercontent.com/1385443/208498335-6b5f5695-aecc-40a1-b05d-97451fde48da.png)

gstrauss commented:
Once it started, it took 3 seconds.  Before it started, it sat almost 30 mins in the queue

Anyway, I made the change you requested.

PromoFaux reviewed (approved):
Looks good to me. It surprises me that it's taken until php 8 to have handy functions such as `str_ends_with`, or even `str_contains`

I suppose after seeing [this SO answer](https://stackoverflow.com/a/4366748)  we _could_ do something like:

```php
if ($_SERVER['REQUEST_METHOD'] === 'POST' && strpos($_SERVER['SCRIPT_NAME'], '/login.php) == false) {
     header('Location: '.$redirect_url);
     exit;
}
```

But really that's just nitpicking on my part. Happy with code as is
